test/SAMtags/parse_mm.pl

```diff
@@ -0,0 +1,107 @@
+# 
+#    2. Redistributions in binary form must reproduce the above copyright notice,
+# this list of conditions and the following disclaimer in the documentation
+# and/or other materials provided with the distribution.
+# 
+#    3. Neither the names Genome Research Ltd and Wellcome Trust Sanger
+# Institute nor the names of its contributors may be used to endorse or promote
+# products derived from this software without specific prior written permission.
+# 
+# THIS SOFTWARE IS PROVIDED BY GENOME RESEARCH LTD AND CONTRIBUTORS "AS IS" AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL GENOME RESEARCH LTD OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use strict;
+
+# Complement a sequence
+sub comp {
+    my ($seq)=@_;
+    $seq =~ tr/ACGTUMRWSYKVHDBN/TGCAAKYWSRMBDHVN/;
+    return $seq;
+}
+
+# Reverse complement a sequence
+sub rc {
+    my ($seq)=@_;
+    $seq = reverse($seq);
+    return comp($seq);
+}
+
+my $nseq = 0;
+while (<>) {
+    chomp($_);                                 # trim newline
+    next if /^\@/;
+
+    my ($qname,$flag,$rname,$pos,$mapq,$cigar,$rnext,
+	$pnext,$tlen,$seq,$qual,@aux) = split("\t",$_);
+    my $dir = $flag & 0x10 ? "-" : "+";
+    my ($mod_str)="@aux"=~m/MM:Z:(\S+)/;
+    my ($mod_prob)="@aux"=~m/ML:B:C,(\S+)/;
+
+    # All counting is from 5' end irrespective of BAM FLAG 0x10
+    $seq = rc($seq) if ($dir eq "-");
+
+    my @seq = split("", $seq);                 # array of seq bases
+    my @seqt = @seq;                           # orientation as shown in SAM
+    my @seqb = split("", comp($seq));          # plus a complemented copy
+
+    $mod_str =~ s/;$//;                        # trim last ; to aid split()
+    my @mods = split(";", $mod_str);
+    my @probs = split(",", $mod_prob);
+    my $pnum = 0;
+
+    print "\n" if $nseq++ > 0;
+    foreach (@mods) {
+	my ($base, $strand, $types, $pos) = $_ =~ m/([A-Z])([-+])([^,]+),(.*)/;
+
+	my $i = 0; # I^{th} bosition in sequence
+	foreach my $delta (split(",", $pos)) {
+	    # Skip $delta occurences of $base
+	    do {
+		$delta-- if ($base eq "N" || $base eq $seq[$i]);
+		$i++;
+	    } while ($delta >= 0);
+	    $i--;
+
+	    # TypePercent combo
+	    my $type_perc = "";
+	    foreach ($types =~ m/(\d+|.)/) {
+		if (/\d/) {
+		    # Avoid ChEBI numbers running into quality values
+		    $type_perc .= "($_)" . int($probs[$pnum++]/255.0*100);
```

Comment:
As per-comment above on the definition of the integer encodings, should this formula be `int(($probs[$pnum++]+0.5)/256.0*100)`.

Reply:
The +0.5 only works if the definition of the integer encoding has an explicit floor.  I don't really like this as it feels more natural (and accurate) when quantising to round to the closest value rather than to floor.  This does mean 0 is absolute certainty of it being incorrect, but that's not a bad thing and I don't see why we'd want to emit such low confidences anyway.  (For what it's worth the same applies to the phred scale and we permit zero there too.)

Is there a specific reason why you want it to floor?

Reply:
To my mind, there is a distinction between Phred and this integer encoding. Phred scaling is a mathematical formula to which rounding is applied in order to report the values as integers (though this rounding is somewhat separated from the definition of the phred score). The integer encoding here maps the continuous range 0-1 to the distinct values 0-255. Since this is simply a mapping and not a formula per se it seems most natural that the mapping is over equally sized portions of the input range (in fact most all interpretations of this encoding do this for all but the first and last intervals). I think the use of +0.5 followed by floor here enforces that the range of the first and last encoded values represent the same portion of the input range as the rest of the values, which seems the most logical interpretation to me. This does mean that conversion back to probability values is approximate/lossy, but that is the nature of compression and this integer encoding specifically.

Reply:
Both are mathematical formulae IMO, but you make a point.  Phred is used only for the most likely event so the lowest values make no sense anyway and won't be used, while the upper values essentially have no defined cap, but become inappropriate due to rarity.    I do see your point on range remapping that all values should be an equal proportion and I didn't think much about the implications of first and last "bins".

Apologies for getting there somewhat slowly on this one!  I haven't been thinking on all cylinders apparently. :)

I think we want a clear English way to describe this as well as a simple formula to back it up.  As you describe it is reasonable e.g. "The continuous probability range 0.0 to 1.0 is remapped in equally sized portions to the discrete integers 0 to 255 inclusive.  Thus the probability range corresponding to integer value `N` is `N/256` to `(N+1)/256`.

The example perl code isn't a normative part of the spec, but agreed with the above definition an additional 0.5 in there makes sense (on the assumption that no knowledge of the actual distribution means the mid-point of each bin is better than the lower boundary, but it's likely not significant enough to matter much either way).

Thanks for your patience.  We have a GA4GH File Formats meeting tomorrow so the other spec maintainers hopefully can comment on this too.

Reply:
I think we're all thinking with a couple cylinders misfiring these days! I'm sure my attempts to explain this fall under the same category.

I think your concise English explanation of the mapping is great. For the formulation with the 0.5, I previously had written a form more in line with the text explanation, but the formula with the 0.5 seemed more concise programatically.

In any case, I'm glad we've come to an understanding of the issue. I'm happy with however you choose to handle the description of this at this point. Thanks again for all the effort on this.

On another note, we should also have a working version of this output in nanopore basecalling software in the coming days for users to start playing around with!

Reply:
That's great to hear.

I've also been working with base modifications to write an interface for querying them within htslib.  In doing so I found a big flaw in my perl example code so it's been doubly useful.